Moved storage size assert to TypeChecker from DeclarationTypeChecker

hrkrshnn opened the pull request:
Replaces: https://github.com/ethereum/solidity/pull/12163
Fixes: https://github.com/ethereum/solidity/issues/12159

hrkrshnn committed:
Moved storage size assert to TypeChecker from DeclarationTypeChecker

hrkrshnn commented on Changelog.md:
```diff
@@ -20,6 +20,7 @@ Bugfixes:
  * Commandline Interface: Disallow ``--error-recovery`` option outside of the compiler mode.
  * SMTChecker: Fix internal error in magic type access (``block``, ``msg``, ``tx``).
  * TypeChecker: Fix internal error when using user defined value types in public library functions.
+ * TypeChecker: Fix internal error when using arrays with user defined value types before declaration.
```
Added this changelog.

cameel replied:
Looks like it's not just arrays. I just found a new failing case (and it fails even with this fix):
```solidity
struct S { U u; }
contract C { S s; }
type U is address;
```
```
Internal compiler error:
/solidity/libsolidity/ast/Types.cpp(2538): Throw in function const solidity::frontend::Type& solidity::frontend::UserDefinedValueType::underlyingType() const
Dynamic exception type: boost::wrapexcept<solidity::langutil::InternalCompilerError>
std::exception::what: Solidity assertion failed
[solidity::util::tag_comment*] = Solidity assertion failed
```

hrkrshnn replied:
Will take a look now. Thanks for noticing!

hrkrshnn replied:
@bshastry Any idea why the fuzzer missed this?

hrkrshnn replied:
This basically opened a bag of bugs. Basically any call to `underlyingType()` from `DeclarationTypeChecker` can be converted into an ICE. Will fix all of them.

hrkrshnn replied:
Wasn't as bad as I thought, other than `canBeStored` (from the struct example), only `isValueType()` will cause issues:
```solidity
MyInt constant x;
type MyInt is int;
```

bshastry replied:
> @bshastry Any idea why the fuzzer missed this?

Because, the grammar-based generator does not do a second pass over type declarations i.e., forward references are not created.

chriseth replied:
@hrkrshnn are the things you mention (`canBeStored`...) already in this PR? Could not find any changes.

hrkrshnn replied:
Not yet, moving https://github.com/ethereum/solidity/blob/develop/libsolidity/analysis/DeclarationTypeChecker.cpp#L442 to TypeChecker created additional issues in other tests. Will keep this as draft for now.

hrkrshnn committed:
Moved a check related to constants to TypeChecker

And added a proper error message when constant types containing (nested) mapping types are used.

hrkrshnn committed:
Moved a canBeStored assert for struct members to TypeChecker

This is to avoid a assert from failing for forward declared user defined value types.

hrkrshnn commented on libsolidity/analysis/DeclarationTypeChecker.cpp:
```diff
@@ -437,14 +437,16 @@ void DeclarationTypeChecker::endVisit(VariableDeclaration const& _variable)
+		m_errorReporter.fatalDeclarationError(
+			3530_error,
+			_variable.location(),
+			"The type contains a (nested) mapping and therefore cannot be a constant."
```
See existing tests with this message. They were throwing, without this additional check.
